### PR TITLE
Allow to pick which Resources will be made unique

### LINKED ISSUE #7009: Allow users to pick which Resources will be made unique when using "Make Unique (Recursive)"
### Describe the project you are working on

N/A (Based on feedback from other contributors and artists)

### Describe the problem or limitation you are having in your project

The current behavior of "Make Unique (Recursive)" can be problematic in situations which involve a large amount of Resources, as it'll make all resources unique, without allowing to specify exceptions.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

We should allow users to pick which resources they want or don't want to make unique, instead of making *all* of the nested Resources unique.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Ideally when selecting "Male Unique (Recursive)" a dialog would pop-up, which would allow the user to pick (in a Tree like fashion) which Resources they *don't* want to make unique.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It must be a core change.

### Is there a reason why this should be core and not an add-on in the asset library?

It can not be an add-on.

### PR DESCRIPTION
**Pull request by KoBeWi**
Closes https://github.com/godotengine/godot-proposals/issues/7009

https://github.com/godotengine/godot/assets/2223172/c69c5ae5-7c2e-4fe3-b5bb-383f3e060d72

"Make Unique (Recursive)" option will now open a popup which allows you to select which resources will be made unique. By default all resources are selected, except Image, Mesh and Shader. (more exceptions can be added, feel free to suggest them)

### CODE REVIEW

**Review by groud — APPROVED** [review]
You are too fast! From a usability standpoint, looks quite good to me.

Maybe the only thing I would add is the disabling of the menu option if the resource has no sub-resource, but that can be done in another PR.

**Comment by Calinou** [comment]
> By default all resources are selected, except Image, Mesh and Shader. (more exceptions can be added, feel free to suggest them)

Should FontData be added?

**Comment by YuriSizov** [comment]
> Hmm, not sure I understand. I've only checked on 4.0.3 where the entry is not hidden when the resource has no sub-resource. Maybe that was changed though ?

No, it was implemented with a check, but perhaps the check is faulty?

https://github.com/godotengine/godot/blob/e23f6aa75c46be7419a8f8e7f52821a5f1f8a1be/editor/editor_resource_picker.cpp#L221-L230

**Comment by KoBeWi** [comment]
The check only checks if the property exists, disregarding whether it was assigned. So it regards `null` as existing sub-resource. I can fix that.

**Review by YuriSizov — CHANGES_REQUESTED** [review]
Implementation looks fine (a couple nitpicks though), but usability needs to be improved a bit.

Currently, it's unclear which resources you are selecting for two reasons:
- We don't display file paths for external resources; we either display types or types + names;
- We don't display which field each resource belongs to.

godot windows editor dev x86_64_2023-07-14_20-59-36

Here's what I think it could look instead:

godot windows editor dev x86_64_2023-07-14_21-10-08

Here's a potential diff that you can use:

```diff
diff --git a/editor/editor_resource_picker.cpp b/editor/editor_resource_picker.cpp
index bfeff41a4a..da78566848 100644
--- a/editor/editor_resource_picker.cpp
+++ b/editor/editor_resource_picker.cpp
@@ -378,6 +378,7 @@ void EditorResourcePicker::_edit_menu_cbk(int p_which) {
 
 				duplicate_resources_tree = memnew(Tree);
 				vb->add_child(duplicate_resources_tree);
+				duplicate_resources_tree->set_columns(2);
 				duplicate_resources_tree->set_v_size_flags(SIZE_EXPAND_FILL);
 			}
 
@@ -386,7 +387,7 @@ void EditorResourcePicker::_edit_menu_cbk(int p_which) {
 			_gather_resources_to_duplicate(edited_resource, root);
 
 			duplicate_resources_dialog->reset_size();
-			duplicate_resources_dialog->popup_centered(Vector2(300, 400) * EDSCALE);
+			duplicate_resources_dialog->popup_centered(Vector2(500, 400) * EDSCALE);
 
 			Ref<Resource> unique_resource = edited_resource->duplicate(true);
 		} break;
@@ -946,21 +947,31 @@ void EditorResourcePicker::_ensure_resource_menu() {
 	edit_menu->connect("popup_hide", callable_mp((BaseButton *)edit_button, &BaseButton::set_pressed).bind(false));
 }
 
-void EditorResourcePicker::_gather_resources_to_duplicate(const Ref<Resource> p_resource, TreeItem *p_item) const {
+void EditorResourcePicker::_gather_resources_to_duplicate(const Ref<Resource> p_resource, TreeItem *p_item, String p_resource_name) const {
 	p_item->set_cell_mode(0, TreeItem::CELL_MODE_CHECK);
 
-	const String res_name = p_resource->get_name();
+	String res_name = p_resource->get_name();
+	if (res_name.is_empty() && p_resource->get_path().is_resource_file()) {
+		res_name = p_resource->get_path().get_file();
+	}
+
 	if (res_name.is_empty()) {
 		p_item->set_text(0, p_resource->get_class());
 	} else {
 		p_item->set_text(0, vformat("%s (%s)", p_resource->get_class(), res_name));
 	}
+
 	p_item->set_icon(0, EditorNode::get_singleton()->get_object_icon(p_resource.ptr()));
 	p_item->set_editable(0, true);
+
 	Array meta;
 	meta.append(p_resource);
 	p_item->set_metadata(0, meta);
 
+	if (!p_resource_name.is_empty()) {
+		p_item->set_text(1, p_resource_name);
+	}
+
 	if (p_resource->get_class() != "Image" && p_resource->get_class() != "Shader" && p_resource->get_class() != "Mesh" && p_resource->get_class() != "FontFile") {
 		// Automatically select resource, unless it's something that shouldn't be duplicated.
 		p_item->set_checked(0, true);
@@ -980,7 +991,7 @@ void EditorResourcePicker::_gather_resources_to_duplicate(const Ref<Resource> p_
 		}
 
 		TreeItem *child = p_item->create_child();
-		_gather_resources_to_duplicate(res, child);
+		_gather_resources_to_duplicate(res, child, E.name);
 
 		meta = child->get_metadata(0);
 		// Remember property name.
diff --git a/editor/editor_resource_picker.h b/editor/editor_resource_picker.h
index 43b6bd7036..897312de7b 100644
--- a/editor/editor_resource_picker.h
+++ b/editor/editor_resource_picker.h
@@ -105,7 +105,7 @@ class EditorResourcePicker : public HBoxContainer {
 	void drop_data_fw(const Point2 &p_point, const Variant &p_data, Control *p_from);
 
 	void _ensure_resource_menu();
-	void _gather_resources_to_duplicate(const Ref<Resource> p_resource, TreeItem *p_item) const;
+	void _gather_resources_to_duplicate(const Ref<Resource> p_resource, TreeItem *p_item, const String p_property_name = "") const;
 	void _duplicate_selected_resources();
 
 protected:
```

File: editor/editor_resource_picker.cpp [hunk 1]
```diff
@@ -920,6 +946,78 @@ void EditorResourcePicker::_ensure_resource_menu() {
...
+void EditorResourcePicker::_duplicate_selected_resources() {
+	for (TreeItem *item = duplicate_resources_tree->get_root(); item; item = item->get_next_in_tree()) {
+		if (!item->is_checked(0)) {
+			continue;
+		}
+
+		Array meta = item->get_metadata(0);
+		Ref<Resource> res = meta[0];
+		Ref<Resource> unique_resource = res->duplicate();
+		ERR_FAIL_COND(unique_resource.is_null());
```
[the remark this answers is not in the document]
Comment (KoBeWi) [reply editor/editor_resource_picker.cpp:1006]: The code is copied from another duplication method.
I looked at the implementation and `duplicate()` may fail when the class can't be instantiated (and possibly in other resource-specific cases, as the method is virtual).

File: editor/editor_resource_picker.cpp [hunk 2]
```diff
@@ -362,12 +364,31 @@ void EditorResourcePicker::_edit_menu_cbk(int p_which) {
...
+			}
+
+			duplicate_resources_tree->clear();
+			TreeItem *root = duplicate_resources_tree->create_item();
+			_gather_resources_to_duplicate(edited_resource, root);
+
+			duplicate_resources_dialog->reset_size();
+			duplicate_resources_dialog->popup_centered(Vector2(300, 400) * EDSCALE);
+
+			Ref<Resource> unique_resource = edited_resource->duplicate(true);
```
[the remark this answers is not in the document]
Comment (KoBeWi) [reply editor/editor_resource_picker.cpp:391]: Leftover .-.

**Comment by KoBeWi** [comment]
@YuriSizov Thanks, I used your patch and added you as co-author.

File: editor/editor_resource_picker.cpp [hunk 3]
```diff
@@ -922,6 +947,89 @@ void EditorResourcePicker::_ensure_resource_menu() {
 	edit_menu->connect("popup_hide", callable_mp((BaseButton *)edit_button, &BaseButton::set_pressed).bind(false));
 }
 
+void EditorResourcePicker::_gather_resources_to_duplicate(const Ref<Resource> p_resource, TreeItem *p_item, const String &p_property_name) const {
+	p_item->set_cell_mode(0, TreeItem::CELL_MODE_CHECK);
+
+	String res_name = p_resource->get_name();
+	if (res_name.is_empty() && !p_resource->is_built_in()) {
```
Comment (YuriSizov) [inline editor/editor_resource_picker.cpp:954]: Note that the check I've suggested originally follows a similar check used in a few places of this file, including a similar resolution for the `assign_button`. But I suppose this should also work.

**Review by YuriSizov — APPROVED** [review]

**Comment by YuriSizov** [comment]
Thanks!

**Pull request merged**